### PR TITLE
Updating Roslyn to consume the Preview4 VSSDK packages.

### PR DESCRIPTION
**Pull request by tannergooding**
FYI. @dotnet/roslyn-infrastructure, @mavasani, @jasonmalinowski 

This is required for Roslyn to build ontop of a Willow install.

### CODE REVIEW

File: build/Targets/Settings.targets [hunk 1]
```diff
@@ -132,13 +132,13 @@
   <Choose>
     <When Condition="'$(VisualStudioVersion)' == '15.0'">
       <PropertyGroup>
-        <VisualStudioBuildToolsNuGetPackagePath>$(NuGetPackageRoot)\Microsoft.VSSDK.BuildTools\15.0.25201-dev15preview2</VisualStudioBuildToolsNuGetPackagePath>
+        <VisualStudioBuildToolsNuGetPackagePath>$(NuGetPackageRoot)\Microsoft.VSSDK.BuildTools\15.0.25604-Preview4</VisualStudioBuildToolsNuGetPackagePath>
```
Comment (jaredpar) [inline build/Targets/Settings.targets]: Looks like I need a way to output the version number for fixed packages as well.

Comment (tannergooding) [inline build/Targets/Settings.targets]: ?

**Comment by mavasani** [comment]
Also tagging @genlu 

👍

**Comment by tannergooding** [comment]
Rebased onto head, I was some commits behind.

**Comment by tannergooding** [comment]
It should probably be noted that this does not allow you to just call `cibuild.cmd` and build on Willow.

There is still a requirement to add a registry key (so NuGet restore can locate MSBuild 15.0 and parse the projects correctly) and a requirement to set `VS150COMNTOOLS`, so `SetDevCommandPrompt` correctly launches the Developer Command Prompt to configure the build environment.

Long term, NuGet needs to support Willow based installs and our build script needs to use the Managed APIs to locate an appropriate VS install (and launch that install's copy of `VsDevCmd.bat`)

**Comment by jasonmalinowski** [comment]
:+1:

**Comment by tannergooding** [comment]
Hmm, that 'ReadInfoFromVsixManifest' bug is exactly what this is supposed to prevent -.-

**Comment by tannergooding** [comment]
So the Preview4 SDK requires a Willow install.

I am going to (tomorrow) add a switch that will use the old 15 VSSDK on Preview 3 and the new one on Willow installs.

File: build/Targets/Settings.targets (hunk 1 above)
Comment (jaredpar) [inline build/Targets/Settings.targets]: Right now packages marked as "fixed" in RepoData.json can't be entries in Dependencies.props.  Hence you have to hard code a version here vs being able to say `$(MicrosoftVSSDKBuildToolsVersion)`

**Comment by tannergooding** [comment]
Updated. This now queries for `$(registry:HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\VisualStudio\$(VisualStudioVersion)@InstallDir)` (as well as the `WOW6432Node` version) and assumes we are a willow installation if it cannot be found.

**Comment by tannergooding** [comment]
Fixed again to only query the registry if OS is `Windows_NT`.

File: build/Targets/Settings.targets [hunk 2]
```diff
@@ -128,17 +128,31 @@
       </PropertyGroup>
     </Otherwise>
   </Choose>
+  
+  <!-- This is a really hacky way to detect whether we are on a legacy or a willow based VS install.
+       Basically, we check for a registry key that will only exist in legacy VS installs, and assume
+       we are a willow based installation if our VSVersion is 15.0 and the registry key doesn't exist. -->
```
Comment (mavasani) [inline build/Targets/Settings.targets]: Another alternative to detect: VsDevCmd.bat in Willow install sets the environment variable "VSCMD_VER", the full VS version. This isn't set with Dev15 preview3 command prompt.

**Comment by mavasani** [comment]
👍

File: build/Targets/Settings.targets (hunk 2 above)
Comment (tannergooding) [inline build/Targets/Settings.targets]: I thought about that, but we need to work if the user explicitly calls MSBuild as well (for example, outside of the Developer Command Prompt).

**Comment by tannergooding** [comment]
Failures were because `VisualStudio.Next` was implicitly pulling in `Newtonsoft.Json`: `6.0.6`, when everything else referenced `8.0.3`. I gave an explicit reference to resolve the conflict.

**Comment by genlu** [comment]
:+1:

**Comment by tannergooding** [comment]
@jasonmalinowski, could I get explicit sign-off from you, now that everything is working as intended?

**Comment by jasonmalinowski** [comment]
I have some suggested cleanup, but this looks functionally correct. :+1:. No qualms about it being merged to unblock.

File: build/Targets/Settings.targets [hunk 3]
```diff
@@ -128,17 +128,31 @@
...
+  <PropertyGroup>
+    <VisualStudioInstallDirectoryFromRegistry Condition="'$(OS)' == 'Windows_NT'">$(registry:HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\VisualStudio\15.0@InstallDir)</VisualStudioInstallDirectoryFromRegistry>
```
Comment (tannergooding) [inline build/Targets/Settings.targets]: I'll fix it up (although I kind of like the name `JasonMalinowskiSaysDoNotUseThis` 😄)

**Comment by Pilchie** [comment]
Didn't @jaredpar remove the use of `Newtonsoft.Json`?

**Comment by jaredpar** [comment]
I removed one incorrect usage in workspaces. I standardized all the rest on 8.0.3 which is the version Visual Studio deploys.

Get Outlook for iOS<https://aka.ms/o0ukef>




On Thu, Aug 11, 2016 at 4:13 PM -0700, "Kevin Pilch-Bisson" <notifications@github.com<mailto:notifications@github.com>> wrote:


Didn't @jaredpar<https://github.com/jaredpar> remove the use of Newtonsoft.Json?

-

File: build/ToolsetPackages/dev15.project.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
     "dependencies": {
-        "Microsoft.VSSDK.BuildTools": "15.0.25201-Dev15Preview2"
+        "Microsoft.VSSDK.BuildTools": "15.0.25201-dev15preview2"
```
Comment (jasonmalinowski) [inline build/ToolsetPackages/dev15.project.json]: Why the casing change?

**Comment by jasonmalinowski** [comment]
@tannergooding not sure if you wanted another :+1:, but :+1:.

File: build/ToolsetPackages/dev15.project.json (hunk 4 above)
Comment (tannergooding) [inline build/ToolsetPackages/dev15.project.json]: The package itself has the lower casing, I was just making it match.

**Pull request merged**